File: C_plus_plus/Minimum_Depth_of_Binary_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode *root) {
        queue<pair<TreeNode *,int>> que;
        if(!root) return 0;
        que.push(make_pair(root,1));
        while(!que.empty()){
            pair<TreeNode *,int> p = que.front();
            que.pop();
            if(!p.first->left&&!p.first->right) 
                return p.second;
            if(p.first->left) que.push(make_pair(p.first->left,p.second+1));
            if(p.first->right) que.push(make_pair(p.first->right,p.second+1));
        }
    }
};

File: java/Minimum_Depth_of_Binary_Tree.java
// BFS queue
// Time is O(n), space is O(n).

public class Solution {
    public int minDepth(TreeNode root) {
        int depth = 0;
        Queue<TreeNode> que = new LinkedList<TreeNode>();
        if (root !=null) que.offer(root);
        while (que.peek() != null) {
            depth++;
            for (int n = que.size(); n > 0 ; --n) {
                TreeNode node = que.poll();
                if (node.left == null && node.right == null)
                    return depth;
                if (node.left != null) que.offer(node.left);
                if (node.right != null) que.offer(node.right);
            }
        }
        return depth;
    }
}

// DFS recursive
// Time and space are both O(n).

public class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return 1;
        if (root.left == null) return minDepth(root.right) + 1;
        if (root.right == null) return minDepth(root.left) + 1;
        return Math.min(minDepth(root.left),minDepth(root.right)) + 1;
    }
}
